### PR TITLE
+13 passing node:vm tests

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

This brings our node:vm implementation closer to Node's.

Fixes https://github.com/oven-sh/bun/issues/11797
Fixes https://github.com/oven-sh/bun/issues/4205
Fixes https://github.com/oven-sh/bun/issues/5658
Properly fixes https://github.com/oven-sh/bun/issues/5847 (happy-dom can delete their workaround)
Fixes https://github.com/oven-sh/bun/issues/6594

Probably fixes (unconfirmed)
- https://github.com/oven-sh/bun/issues/10282
- https://github.com/oven-sh/bun/issues/6044

### How did you verify your code works?

Node tests!

### CODE REVIEW

**Comment by robobun** [comment]
<div><sup>Updated 6:21 PM PT - Dec 18th, 2024</sup></div>

@Jarred-Sumner, your commit 8cee321facc55c3cd1b3c031d9e3ba48b8745487 is building: [`#8192`](https://buildkite.com/bun/bun/builds/8192)

File: src/bun.js/bindings/NodeVM.cpp [hunk 1]
```diff
@@ -28,9 +35,251 @@
...
+    }
+
+    slot.setThisValue(sandbox);
+
+    if (!sandbox->methodTable()->put(sandbox, globalObject, propertyName, value, slot)) {
+        return false;
+    }
+    RETURN_IF_EXCEPTION(scope, false);
+
+    slot.setThisValue(sandbox);
```
Comment (dylan-conway) [inline src/bun.js/bindings/NodeVM.cpp:203]: can we move this inside the `if` block?

File: src/bun.js/bindings/NodeVM.cpp [hunk 2]
```diff
@@ -189,114 +424,137 @@ JSC_DEFINE_HOST_FUNCTION(scriptCreateCachedData, (JSGlobalObject * globalObject,
...
-    JSValue result = vm.interpreter.executeEval(executable, targetContext, contextScope);
-    if (UNLIKELY(catchScope.exception())) {
-        auto returnedException = catchScope.exception();
-        catchScope.clearException();
-        JSC::throwException(globalObject, throwScope, returnedException);
-    }
-
-    RETURN_IF_EXCEPTION(throwScope, {});
+    if (exception)
+        throwException(globalObject, scope, exception);
```
Comment (dylan-conway) [inline src/bun.js/bindings/NodeVM.cpp:557]: not important, but should zero be returned here

**Review by dylan-conway — APPROVED** [review]

**Comment by Spixmaster** [comment]
@Jarred-Sumner Thank you very much!

**Pull request merged**